### ISSUE TITLE
Flashes are not getting reflashed on 409

### ISSUE THREAD

**Issue by totaltrash** [devato/inertia_phoenix#16]
https://inertiajs.com/the-protocol#asset-versioning

> Finally, in the event that flash session data exists when a 409 Conflict response occurs, the server will automatically reflash this data.

**Comment by tmartin314** [comment]
Thanks @totaltrash .

**Comment by bigx333** [comment]
Fetch_flash does basically the same for redirects https://github.com/phoenixframework/phoenix/blob/master/lib/phoenix/controller.ex#L1296

Can use the same logic

**Comment by totaltrash** [comment]
Thanks @bigx333 . Just having a look at this now, but wondering why [controller.ex](https://github.com/phoenixframework/phoenix/blob/master/lib/phoenix/controller.ex) overrides all of the out-of-the-box [phoenix flash handling](https://hexdocs.pm/phoenix/controllers.html#flash-messages)?

A simpler solution be to store flash messages as normal phoenix flashes. Then at the time of preparing the inertia response, pull those flashes out of the session and merging them into the page data in the same way the shared data does.

**Comment by totaltrash** [comment]
ugh, sorry, it's still early here. ignore my previous comment!

**Comment by totaltrash** [comment]
I've had a nightmare with this. I can't write a failing test (I think it needs an integration test to replicate it), and I can't write an implementation... my elixir chops are not quite there.

Looking at how Phoenix handles flashes, [as per your link](https://github.com/phoenixframework/phoenix/blob/master/lib/phoenix/controller.ex#L1296) @bigx333 , the flashes get deleted from the session before sending the response (unless a redirect). Se we'll need to store the flashes under a different session key, and merge them in upon the next request after a 409? Or something else?

**Comment by tmartin314** [comment]
It looks like Jose just updated that plug in the new version of Phoenix:  

https://github.com/phoenixframework/phoenix/commit/9fe4c018d346d3528adf432227a99cbd48965070

**Comment by bigx333** [comment]
> I've had a nightmare with this. I can't write a failing test (I think it needs an integration test to replicate it), and I can't write an implementation... my elixir chops are not quite there.
> 
> Looking at how Phoenix handles flashes, [as per your link](https://github.com/phoenixframework/phoenix/blob/master/lib/phoenix/controller.ex#L1296) @bigx333 , the flashes get deleted from the session before sending the response (unless a redirect). Se we'll need to store the flashes under a different session key, and merge them in upon the next request after a 409? Or something else?

I didn't really have time to look into this but fetch_flash removes the session after it saves the flash on conn.private.phenix_flash, you should be able to just use put_session to persist the flash just like fetch_flash does and fetch_flash will automatically repopulate the conn.private on the next request.

Something like this should work (untested)

```elixir
|> check_flash()
|> send_resp(:conflict, "")
```

```elixir
def check_flash(conn) do
    flash = conn.private.phoenix_flash
    flash_size = map_size(flash)

    if flash_size > 0 do
      put_session(conn, "phoenix_flash", flash)
    else
      conn
    end
  end
```

Regarding the test, I'm not quite sure, normally on a phoenix test using get("/") would recycle the connection and set the session based on the cookies but here we don't have a router so this needs to be done manually.

@tmartin314 Any input on writing a test for this?

**Comment by tmartin314** [comment]
Might be worth while to refactor the tests to use a dummy router and controller so we can test requests regularly.   The controller test we have started out pretty simple and needs to cover many different scenarios now.

I'll start re-working the `ConnCase`.

**Comment by tmartin314** [comment]
@totaltrash @bigx333 I was able to refactor the tests in 0.2.4.  I'll take a stab at this issue as well.  Thanks again for all your help.

**Comment by totaltrash** [comment]
G'day, I've pulled the latest changes including #24 into my test app. The flashes aren't getting reflashed upon a 409.

> fetch_flash removes the session after it saves the flash on conn.private.phenix_flash, you should be able to just use put_session to persist the flash just like fetch_flash does and fetch_flash will automatically repopulate conn.private on the next request

I think the issue is that it clears the session _before sending the response_ using `register_before_send`.

So the timing is:

1. `fetch_flash` plug is called by the router, loading the flashes out of the session, and a callback is registered with [`register_before_send`](https://github.com/phoenixframework/phoenix/blob/master/lib/phoenix/controller.ex#L1349)
2. `InertiaPhoenix.plug` is called, an invalid inertia version is identified, and any existing flashes are written back to the session.
3. The `register_before_send` is called, which [deletes the session](https://github.com/phoenixframework/phoenix/blob/master/lib/phoenix/controller.ex#L1359) (as the status in not a redirect) including the reflashed flashes

**Comment by tmartin314** [comment]
@totaltrash Ok thanks.  Can you create a new issue with some demo code on how to reproduce the issue?

**Comment by tmartin314** [comment]
The good news is that the tests should be easier to work with now

**Comment by tmartin314** [comment]
@totaltrash It looks like @bigx333 tackled this issue.  It's merged into v0.2.6

**Comment by totaltrash** [comment]
Yep, nailed it, nice @bigx333
